File: DebugClients/Python3/AsyncIO.py
# -*- coding: utf-8 -*-

# Copyright (c) 2009 - 2015 Detlev Offenbach <detlev@die-offenbachs.de>
#

"""
Module implementing a base class of an asynchronous interface for the debugger.
"""


class AsyncIO(object):
    """
    Class implementing asynchronous reading and writing.
    """
    def __init__(self):
        """
        Constructor
        """
        # There is no connection yet.
        self.disconnect()

    def disconnect(self):
        """
        Public method to disconnect any current connection.
        """
        self.readfd = None
        self.writefd = None

    def setDescriptors(self, rfd, wfd):
        """
        Public method called to set the descriptors for the connection.
        
        @param rfd file descriptor of the input file (int)
        @param wfd file descriptor of the output file (int)
        """
        self.rbuf = ''
        self.readfd = rfd

        self.wbuf = ''
        self.writefd = wfd

    def readReady(self, fd):
        """
        Public method called when there is data ready to be read.
        
        @param fd file descriptor of the file that has data to be read (int)
        """
        try:
            got = self.readfd.readline_p()
        except Exception:
            return

        if len(got) == 0:
            self.sessionClose()
            return

        self.rbuf = self.rbuf + got

        # Call handleLine for the line if it is complete.
        eol = self.rbuf.find('\n')

        while eol >= 0:
            s = self.rbuf[:eol + 1]
            self.rbuf = self.rbuf[eol + 1:]
            self.handleLine(s)
            eol = self.rbuf.find('\n')

    def writeReady(self, fd):
        """
        Public method called when we are ready to write data.
        
        @param fd file descriptor of the file that has data to be written (int)
        """
        self.writefd.write(self.wbuf)
        self.writefd.flush()
        self.wbuf = ''

    def write(self, s):
        """
        Public method to write a string.
        
        @param s the data to be written (string)
        """
        self.wbuf = self.wbuf + s

File: DebugClients/Ruby/AsyncIO.rb
# -*- coding: utf-8 -*-

# Copyright (c) 2005 - 2015 Detlev Offenbach <detlev@die-offenbachs.de>
#

=begin edoc
File implementing an asynchronous interface for the debugger.
=end

module AsyncIO
=begin edoc
Module implementing asynchronous reading and writing.
=end
    def initializeAsyncIO
=begin edoc
Function to initialize the module.
=end
        disconnect()
    end
    
    def disconnect
=begin edoc
Function to disconnect any current connection.
=end
        @readfd = nil
        @writefd = nil
    end
    
    def setDescriptors(rfd, wfd)
=begin edoc
Function called to set the descriptors for the connection.

@param fd file descriptor of the input file (int)
@param wfd file descriptor of the output file (int)
=end
        @rbuf = ''
        @readfd = rfd

        @wbuf = ''
        @writefd = wfd
    end
    
    def readReady(fd)
=begin edoc
Function called when there is data ready to be read.

@param fd file descriptor of the file that has data to be read (int)
=end
        begin
            got = @readfd.readline()
        rescue
            return
        end
        
        if got.length == 0
            sessionClose()
            return
        end
        
        @rbuf << got
        
        # Call handleLine for the line if it is complete.
        eol = @rbuf.index("\n")
        
        while eol and eol >= 0
            s = @rbuf[0..eol]
            @rbuf = @rbuf[eol+1..-1]
            handleLine(s)
            eol = @rbuf.index("\n")
        end
    end
    
    def writeReady(fd)
=begin edoc
Function called when we are ready to write data.

@param fd file descriptor of the file that has data to be written (int)
=end
        @writefd.write(@wbuf)
        @writefd.flush()
        @wbuf = ''
    end
    
    def write(s)
=begin edoc
Function to write a string.

@param s the data to be written (string)
=end
        @wbuf << s
    end
end
